app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/TrackListViewModel.kt

```diff
@@ -19,14 +20,47 @@ class TrackListViewModel
 @Inject
 constructor(
     private val spotifyController: SpotifyController,
+    private val appDatabase: AppDatabase,
     private val repository: TrackRepository
 ) : ViewModel() {
 
   private val _uiState = MutableStateFlow(UiState(loading = true))
   val uiState: StateFlow<UiState> = _uiState
+  private var _currentBeaconId: String? = null
+  var currentBeaconId: String?
+    get() = _currentBeaconId
+    set(value) {
+      _currentBeaconId = value
+      loadTracksBasedOnSource()
+    }
 
   private var _searchQuery = MutableStateFlow("")
 
+  fun toggleTrackSource() {
+    _uiState.value = _uiState.value.copy(showRecentlyAdded = !_uiState.value.showRecentlyAdded)
+    loadTracksBasedOnSource()
+  }
+
+  fun loadTracksBasedOnSource() {
+    viewModelScope.launch {
+      _uiState.value = _uiState.value.copy(loading = true)
+      if (_uiState.value.showRecentlyAdded && currentBeaconId != null) {
+        loadRecentlyAddedTracks(currentBeaconId!!)
+      } else {
+        _uiState.value = _uiState.value.copy(tracks = listOf(), loading = false)
+      }
+    }
+  }
+
+  fun loadRecentlyAddedTracks(beaconId: String) {
```

Comment:
Maybe there was a misunderstanding in the requirements here. The way I understood it, the Track list should display the songs added recently to ANY beacon, not just to a specific one...

Reply:
oh okay I see , I will fix that